pyext/src/output.py

```diff
@@ -691,57 +691,119 @@ def write_stats2(self):
+            self.inv_dict = {}
+
+            for k in kkeys:
+                self.inv_dict.update({self.dict[k]: k})
+        else:
+            print("WARNING: statfile v1 is deprecated.  Please convert to statfile v2")
+            self.isstat1 = True
+            self.klist.sort()
+        
 
-            break
         f.close()
 
+
+    def parse_line(self, line):
+        # Parses a line and returns a dictionary of key:value pairs
+        # {key:value, key:value, key:"{v1, v2, v3}", key:value}
+        d={} # output dictionary 
+        # First, remove outer braces and split via double quotes to isolate multi-component values.
+        split=line.strip()[1:-1].split("\"")
+        
+        if len(split)==1:
+            fields = split[0].split(",")   # split via commas to get key:value pair
+            for h in fields[0:-1]: 
+                if h != "":  # For some reason, there is occasionally an empty field. Ignoring these seems to work.
+
+                    # Split fields into key and value elements
+                    kv = h.split(":")
+
+                    # If the key (field 0) is an integer, keep it an integer.
+                    try:
+                        k = int(kv[0].replace(",","").strip())
+                    except:
+                        k = kv[0].replace("\'","").strip()
+
+                    v = kv[1].replace("\'","").strip() # the value is encased in single quote characters, so remove these
+                    d[k] = v
+            return d
+
+
+        for i in range(0,len(split)-1,2):
+            # Each even number of split is a string of "key:value, key:value, key:value"
+            fd = split[i]
+            fields = fd.split(",")   # split via commas to get key:value pair
+            for h in fields[0:-1]: 
+                if h != "":  # For some reason, there is occasionally an empty field. Ignoring these seems to work.
+
+                    # Split fields into key and value elements
+                    kv = h.split(":")
+
+                    # If the key (field 0) is an integer, keep it an integer.
+                    try:
+                        k = int(kv[0].replace(",","").strip())
+                    except:
+                        k = kv[0].replace("\'","").strip()
+
+                    v = kv[1].replace("\'","").strip() # the value is encased in single quote characters, so remove these
+                    d[k] = v
+            # The last field contains the key for the multi-component value in split[i+1]
+            d[fields[-1].split(":")[0].replace("\'","").strip()] = split[i+1]
+
+        return d
+
+
     def get_keys(self):
+        self.klist = [k[1]
+                    for k in sorted(self.dict.items(), key=operator.itemgetter(1))]
         return self.klist
 
     def show_keys(self, ncolumns=2, truncate=65):
         IMP.pmi.tools.print_multicolumn(self.get_keys(), ncolumns, truncate)
 
+    def get_experimental_values(self):
```

Comment:
Need a test to make sure this returns the right thing for a reasonably complex statfile.